src/protocol/message.rs

```diff
@@ -179,3 +180,187 @@ fn checksum(bytes: &[u8]) -> u32 {
+    ping: Filter,
+    getheaders: Filter,
+    getaddr: Filter,
+    getdata: Filter,
+    // todo: inv
+    // todo: getblocks
+    // todo: mempool
+    logging: bool,
+}
+
+impl MessageFilter {
+    /// Constructs a [MessageFilter] which will filter no messages, and with logging disabled.
+    pub fn with_all_disabled() -> Self {
+        use Filter::Disabled;
+
+        Self {
+            ping: Disabled,
+            getheaders: Disabled,
+            getaddr: Disabled,
+            getdata: Disabled,
+
+            logging: false,
+        }
+    }
+
+    /// Constructs a [MessageFilter] which will filter all supported message types, and with logging disabled.
+    pub fn with_all_enabled() -> Self {
+        use Filter::Enabled;
+
+        Self {
+            ping: Enabled,
+            getheaders: Enabled,
+            getaddr: Enabled,
+            getdata: Enabled,
+
+            logging: false,
+        }
+    }
+
+    /// Constructs a [MessageFilter] which will filter and reply to all supported message types, and with logging disabled.
+    pub fn with_all_auto_reply() -> Self {
+        use Filter::AutoReply;
+
+        Self {
+            ping: AutoReply,
+            getheaders: AutoReply,
+            getaddr: AutoReply,
+            getdata: AutoReply,
+
+            logging: false,
+        }
+    }
+
+    /// Enables logging filter events to console
+    pub fn enable_logging(mut self) -> Self {
+        self.logging = true;
+        self
+    }
+
+    /// Sets the [Filter] response for [Message::GetHeaders] messages
+    pub fn with_getheaders_filter(mut self, filter: Filter) -> Self {
+        self.getheaders = filter;
+        self
+    }
+
+    /// Sets the [Filter] response for [Message::GetAddr] messages
+    pub fn with_getaddr_filter(mut self, filter: Filter) -> Self {
+        self.getaddr = filter;
+        self
+    }
+
+    /// Sets the [Filter] response for [Message::GetData] messages
+    pub fn with_getdata_filter(mut self, filter: Filter) -> Self {
+        self.getdata = filter;
+        self
+    }
+
+    /// Sets the [Filter] response for [Message::Ping] messages
+    pub fn with_ping_filter(mut self, filter: Filter) -> Self {
+        self.ping = filter;
+        self
+    }
+
+    // sends an appropriate reply in response to the received message
+    async fn reply(&self, stream: &mut TcpStream, message: Message) -> io::Result<()> {
+        let result = match message {
+            Message::Ping(nonce) => Message::Pong(nonce).write_to_stream(stream).await,
+            Message::GetAddr => Message::Addr(Addr::empty()).write_to_stream(stream).await,
+            Message::GetHeaders(_) => {
+                Message::Headers(Headers::empty())
+                    .write_to_stream(stream)
+                    .await
+            }
+            Message::GetData(inv) => Message::NotFound(inv).write_to_stream(stream).await,
+            _ => unimplemented!(),
+        };
+
+        result
```

Comment:
Suggestion: instead assigning the result, you could use `?` to bubble the error and return `Ok(())` if the execution isn't short-circuited.

Reply:
That's what I get for copy pasting a previous attempt 🤦

I can in-fact simply return the result straight from the match.